test/testother.cpp

```diff
@@ -4100,6 +4101,17 @@ class TestOther : public TestFixture {
         ASSERT_EQUALS("", errout.str());
     }
 
+    void duplicateBranch4() {
+        check("void* f(bool b) {\n"
```

Comment:
The tests here should be unit tests and specifically test the logic in `duplicateBranch` so this check doesn't _really_ belong here. I want this to be clarified/separated somehow. But I am not sure how I'd like to do that.

I do not think of this as a unit test but rather a integration test or something.

Tests should prevent regressions. And I believe that the AST test will primarily test that the real bug is avoided. The motivation for this test (in my head) would be if we by mistake remove the AST test or rewrite it wrongly.. or something like that. Or if we for some reason stops using AST later..

Well I am not sure how to separate integration tests well. Do you have some ideas? If you would add a comment like `// integration test; AST is bad => isSameExpression fails => duplicateBranch FP` then we can later understand that it's an integration test and what potential regressions the test could catch.

But somehow I would like to separate integration tests and unit tests.

This is not the first and only integration test in testrunner but the absolute majority of the tests are unit tests. Somehow it would feel good to have a clear strategy and ensure if all existing integration tests would follow that.

Reply:
As for the value of this test:
I think duplicateBranch tests should verify that duplicateBranch warnings occur only when they should occur without having to know how they're implemented. If we eliminate test cases because we know there are tests of other units which should catch them then we're not black box testing and we link the success of the test to the validity of that assumption.

As for where it belongs:
I agree this is an integration test rather than a unit test but in my experience the distinction can be quite unhelpful and the line is often not clear. So, since you asked, my opinion is that they're not worth separating. I'm happy to enact whatever suggestion you have though. If the best solution is to add the comment you propose I'll happily do that. However, one day we may refactor duplicateBranch so that it no longer calls isSameExpression. At that point this comment would be incorrect but the test case would be no less valid. In other words we're not testing this case to make sure the ast hasn't failed. We're testing it to make sure #9700 doesn't happen again.

I think a unit test for this behavior would require the ability to inject a complete ast into the checker without relying on Tokenizer to generate that ast. I don't think that's very easy to do though and Ultimately I'm not sure it's an improvement (unless we wanted to support multiple ast generators in which case it would be necessary regardless)

Reply:
> At that point this comment would be incorrect but the test case would be no less valid.

yes true!

How do you feel about such alternative approach: I imagine that we could create a script that would query our trac database. It's a sqlite database so that is not rocket science. For "Fixed" "False positive" tickets we could extract the example code from the description if that is enclosed inside `{{{` and `}}}` and preceded by the text "Example code:". The script could execute cppcheck on those code examples. If no warning is shown then the script is happy. If warnings are shown.. well then it should try to check if those match the ticket.. In quite many false positive tickets the message id is added as a keyword. In 9700 it would not be possible to extract the example code though. So well.. that ticket could be tweaked.

Reply:
It's not something I'd want to do as part of this pull request :P. I think the idea has merit but I think it would be better to run the script once to extract test case that you then upload to github. You don't want the outcome of your builds to be dependent on the state of Trac (e.g. someone doesn't structure their Trac ticket in the way you expect and now everyone's builds fail).